### PR TITLE
Add mutipaxos support

### PR DESCRIPTION
**Pull request by rsingh16**

### CODE REVIEW

File: paxos/proposer.cc [hunk 1]
```diff
@@ -1,125 +1,138 @@
...
   bool done = false;
   while (!done) {
     std::string value_for_accept_phase = value;
-    uint32_t num_promises = 0;
-
-    // Perform phase 1 of the paxos operation i.e. find a proposal that will be
-    // accepted by a quorum of acceptors.
     paxos_rpc::PrepareRequest request;
     request.set_index(this->replicated_log_->GetFirstUnchosenIdx());
+    request.set_proposal_number( proposal_number_ );
```
Comment (vramesh7) [inline paxos/proposer.cc:15]: Nit: Extra space

File: paxos/proposer.cc [hunk 2]
```diff
@@ -1,125 +1,149 @@
...
-      this->paxos_node_->CommitOnPeerNodes(peer_unchosen_idx);
-      break;
-    }
+  if (done || nop_paxos_round) {
+    done = true;
+    this->paxos_node_->CommitOnPeerNodes(peer_unchosen_idx);
   }
 }
-}  // namespace witnesskvs::paxos
+}  // namespace witnesskvs::paxos
```
Comment (mmucklo) [inline paxos/proposer.cc:149]: Nit: add newline

File: paxos/proposer.hh [hunk 3]
```diff
@@ -15,23 +15,34 @@ namespace witnesskvs::paxos {
 
 class Proposer {
  private:
+  uint64_t proposal_number_;
```
Comment (vramesh7) [inline paxos/proposer.hh:18]: Why is this field even needed ? The vector of bool should be enough right ?

File: paxos/proposer.hh [hunk 4]
```diff
@@ -15,23 +15,34 @@ namespace witnesskvs::paxos {
...
   uint8_t node_id_;
   int majority_threshold_;
 
   std::shared_ptr<ReplicatedLog> replicated_log_;
   std::shared_ptr<PaxosNode> paxos_node_;
+  std::vector<bool> is_prepare_needed_;
```
Comment (vramesh7) [inline paxos/proposer.hh:24]: I think when we become leader this vector must be initialized for all values to be true

File: paxos/proposer.hh (hunk 3 above)
Comment (rsingh16) [reply paxos/proposer.hh:18]: we need this to save proposal number after first one to use for subsequent accept request.

File: paxos/proposer.hh (hunk 4 above)
Comment (vramesh7) [reply paxos/proposer.hh:24]: Or at least make the proposal number = 0

Comment (rsingh16) [reply paxos/proposer.hh:24]: I think we do, but let me check once.

Comment (rsingh16) [reply paxos/proposer.hh:24]: After a leader change we must make sure the prepares happen
>> We do, I have a unittest for this already where I change the leader and make sure we are doing prepare with new proposal id.

**Comment by rsingh16** [comment]
We do set it 0 in node.cc, when we create proposer object for noop, the proposal_id will get initialized to 0.

File: paxos/proposer.hh (hunk 4 above)
Comment (rsingh16) [reply paxos/proposer.hh:24]: I verified the proposal number 0 on leader change in both ways:
**Method 1, I have added a unittest:**

1. Start with 3 nodes, node 1 is the leader:
     It start with proposal number 0
     Gets a proposal id, after first prepare
     Uses that for it's remaining operations without prepare
2. Node 1 goes down, node 0 becomes leader:
    Node 0 start with proposal number 0
    Gets a proposal id, after first prepare
    Uses that for it's remaining operations without prepare
3. Node 1 comes backup, becomes leader:
    Node 1 start with proposal id 0
    Gets a proposal id after first prepare
    Keeps doing prepare until it fills the holes up, does 5 prepare to catch up
    Once caught up, uses the same proposal id for remaining operations without prepare
4. Node 1 goes down again, node 0 becomes the leader
   Node 0 starts with proposal id 0
   Gets a proposal id after first prepare
   Uses that for it's remaining operations without prep

All of this logic is in the unittest.

**Method 2, manual verification by running the client and server:**

1. Start 3 servers, node 1 is leader
    Same as above
2. Kill node 1, node 0 is leader now
    same as above
3. Bring back node 1, it becomes leader
    same as above, does prep with different proposal id until fills holes and then only accept with same proposal id.
4. Kill node 1 again, node 0 becomes leader
   same as above

File: paxos/proposer.cc [hunk 5]
```diff
@@ -1,125 +1,150 @@
...
+
+}
 
-    // Since a quorum of acceptors responded with an accept for this value,
-    // we can mark this entry as chosen.
-    if (accept_majority_count >= majority_threshold_) {
-      CHECK(!nop_paxos_round)
-          << "NODE: [" << static_cast<uint32_t>(node_id_)
-          << "] NOP round of paxos should never make log commits.";
+void Proposer::AcceptPhase(paxos_rpc::PrepareRequest& request,
```
Comment (vramesh7) [inline paxos/proposer.cc:84]: I don't want to block this review but done should be returned by this function. Instead of a taking it by a reference

Comment (vramesh7) [reply paxos/proposer.cc:84]: Can be done as a follow up or something also

File: paxos/acceptor.cc [hunk 6]
```diff
@@ -70,6 +70,7 @@ Status AcceptorImpl::Prepare(ServerContext* context,
     response->set_accepted_value(entry.accepted_value_);
     response->set_has_accepted_value(true);
   }
+  response->set_max_chosen_index(this->replicated_log_->GetLogEntries().rbegin()->first);
```
Comment (vramesh7) [inline paxos/acceptor.cc:73]: This is not the max chosen idx. This is just the max idx (can be accepted also).
This should be named max_idx_in_log or something

**Review by vramesh7 — APPROVED** [review]

File: paxos/proposer.cc (hunk 5 above)
Comment (rsingh16) [reply paxos/proposer.cc]: fixed

File: paxos/acceptor.cc (hunk 6 above)
Comment (rsingh16) [reply paxos/acceptor.cc]: fixed

**Comment by rsingh16** [comment]
Addressed code review comment, merging.

**Pull request merged**